XMonad/Actions/CycleRecentWS.hs

```diff
@@ -76,56 +80,90 @@ toggleRecentNonEmptyWS :: X ()
-                -> [KeySym]                   -- ^ A list of modifier keys used when invoking this action.
-                                              --   As soon as one of them is released, the final WindowSet is chosen and the action exits.
-                -> KeySym                     -- ^ Key used to preview next WindowSet from the list of generated options
-                -> KeySym                     -- ^ Key used to preview previous WindowSet from the list of generated options.
-                                              --   If it's the same as nextOption key, it is effectively ignored.
+cycleWindowSets :: (WindowSet -> [WorkspaceId]) -- ^ A function used to create a list of workspaces to choose from
+                -> [KeySym]                     -- ^ A list of modifier keys used when invoking this action.
+                                                --   As soon as one of them is released, the final workspace is chosen and the action exits.
+                -> KeySym                       -- ^ Key used to preview next workspace from the list of generated options
+                -> KeySym                       -- ^ Key used to preview previous workspace from the list of generated options.
+                                                --   If it's the same as nextOption key, it is effectively ignored.
                 -> X ()
 cycleWindowSets genOptions mods keyNext keyPrev = do
-  options <- gets $ genOptions . windowset
+  origWSet <- gets windowset
+  let options = genOptions origWSet
   XConf {theRoot = root, display = d} <- ask
   let event = allocaXEvent $ \p -> do
                 maskEvent d (keyPressMask .|. keyReleaseMask) p
                 KeyEvent {ev_event_type = t, ev_keycode = c} <- getEvent p
                 s <- keycodeToKeysym d c 0
                 return (t, s)
-  let setOption n = do windows $ const $ options `cycref` n
-                       (t, s) <- io event
-                       case () of
-                         () | t == keyPress   && s == keyNext  -> setOption (n+1)
-                            | t == keyPress   && s == keyPrev  -> setOption (n-1)
-                            | t == keyRelease && s `elem` mods -> return ()
-                            | otherwise                        -> setOption n
+  let setOption n = do
+        let ws = options `cycref` n
+        (t, s) <- io event
+        if | t == keyPress   && s == keyNext  ->
+               windows (view ws) >> setOption (n + 1)
+           | t == keyPress   && s == keyPrev  ->
+               windows (view ws) >> setOption (n - 1)
+           | t == keyRelease && s `elem` mods -> do
+               foc <- gets $ tag . workspace . current . windowset
+               -- If there are any inconsistencies when restoring, give
+               -- up and just return a no-op.
+               windows $ \w -> maybe w (view foc) (restoreOrder origWSet w)
```

Comment:
Okay I cleaned this up a bit to hopefully make it a little bit more readable.  At least on a single monitor this works now; will check on a mutli-monitor setup once have access to that again.

Reply:
I added a quickcheck property `restoreOrder ss (view t ss) === ss` and it's falsifiable:

https://x0.at/kHw

(The reason I added it was because I wanted to simplify the code, as complex code is more likely to contain bugs… Anyway, I'll continue working on it, so no need to try to fix it now.)